### ISSUE TITLE
Narrow type in negative case when class is declared as final

### ISSUE THREAD

**Issue by ethframe** [microsoft/pyright#2138]
**Is your feature request related to a problem? Please describe.**

I have the following code:

```python
from typing import Union

from typing_extensions import final


@final
class A:
    pass


@final
class B:
    pass


C = Union[A, B]


def foo(c: C) -> None:
    if type(c) is A:
        reveal_type(c)
    else:
        reveal_type(c) 
```

I am trying to narrow the type using `type(x) is T` type guard. However, pyright narrows type only in positive case, so I am forced to do two checks instead of one:

```
$ pyright foo.py
...
  foo.py:21:21 - info: Type of "c" is "A"
  foo.py:23:21 - info: Type of "c" is "A | B"
```

**Describe the solution you'd like**

`type(x) is T` should narrow the type in a negative case when `T` is a class declared as final. 

**Additional context**

This also would be consistent with mypy:

```
$ mypy foo.py
foo.py:21: note: Revealed type is "foo.A"
foo.py:23: note: Revealed type is "foo.B"
```

**Comment by erictraut** [comment]
Thanks for the suggestion. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is fixed in pyright 1.1.160, which I just published. It will also be included in the next release of pylance.
